apps/user-cdc/src/main/java/it/pagopa/selfcare/user/event/repository/UserInstitutionRepository.java

```diff
@@ -0,0 +1,105 @@
+package it.pagopa.selfcare.user.event.repository;
+
+import io.quarkus.mongodb.panache.reactive.ReactivePanacheMongoEntityBase;
+import io.smallrye.mutiny.Uni;
+import it.pagopa.selfcare.onboarding.common.PartyRole;
+import it.pagopa.selfcare.user.event.constant.OnboardedProductState;
+import it.pagopa.selfcare.user.event.entity.OnboardedProduct;
+import it.pagopa.selfcare.user.event.entity.UserInfo;
+import it.pagopa.selfcare.user.event.entity.UserInstitution;
+import it.pagopa.selfcare.user.event.entity.UserInstitutionRole;
+import it.pagopa.selfcare.user.event.mapper.UserMapper;
+import jakarta.enterprise.context.ApplicationScoped;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.*;
+
+
+@Slf4j
+@RequiredArgsConstructor
+@ApplicationScoped
+public class UserInstitutionRepository {
+    private static final List<OnboardedProductState> VALID_PRODUCT_STATE = List.of(OnboardedProductState.ACTIVE, OnboardedProductState.PENDING, OnboardedProductState.TOBEVALIDATED);
+
+    private final UserMapper userMapper;
+
+    public void updateUser(UserInstitution userInstitution) {
+        OnboardedProductState state = retrieveStatusForGivenInstitution(userInstitution.getProducts());
+        UserInfo.findByIdOptional(userInstitution.getUserId())
+                .onItem().transformToUni(opt -> {
+                    if (VALID_PRODUCT_STATE.contains(state)) {
+                        PartyRole role = retrieveRoleForGivenInstitution(userInstitution.getProducts());
+                        return updateOrCreateNewUserInfo(opt, userInstitution, role, state);
+                    } else {
+                        return deleteInstitutionOrAllUserInfo(opt, userInstitution);
+                    }
+                })
+                .subscribe().with(
+                        result -> log.info("UserInfo collection successfully updated"),
+                        failure -> log.error("Error during UserInfo collection updating, message:" + failure.getMessage()));
+
+    }
+
+    private Uni<Void> deleteInstitutionOrAllUserInfo(Optional<ReactivePanacheMongoEntityBase> opt, UserInstitution userInstitution) {
+        return opt.map(UserInfo.class::cast)
+                .map(userInfo -> {
+                    if (userInfo.getInstitutions().stream()
+                            .anyMatch(userInstitutionRole -> userInstitutionRole.getInstitutionId().equalsIgnoreCase(userInstitution.getInstitutionId()))) {
+
+                        List<UserInstitutionRole> roleList = new ArrayList<>(userInfo.getInstitutions());
+                        roleList.removeIf(userInstitutionRole -> userInstitutionRole.getInstitutionId().equalsIgnoreCase(userInstitution.getInstitutionId()));
+                        userInfo.setInstitutions(roleList);
```

Comment:
you could simplify using stream:

List<UserInstitutionRole> roleList = userInfo.getInstitutions().stream().filter(userInstitutionRole -> userInstitutionRole.getInstitutionId().equalsIgnoreCase(userInstitution.getInstitutionId())).collect(Collections.toList())